server/src/main/java/org/opensearch/common/blobstore/stream/read/listener/StreamCompletionListener.java

```diff
@@ -0,0 +1,88 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.common.blobstore.stream.read.listener;
+
+import org.opensearch.common.io.Channels;
+import org.opensearch.common.io.InputStreamContainer;
+import org.opensearch.core.action.ActionListener;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.nio.channels.FileChannel;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardOpenOption;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public class StreamCompletionListener implements ActionListener<Void> {
+    private final int partNumber;
+    private final InputStreamContainer blobPartStreamContainer;
+    private final Path fileLocation;
+    private final AtomicBoolean anyPartStreamFailed;
+    private final FileCompletionListener fileCompletionListener;
+
+    // 8 MB buffer for transfer
+    private static final int BUFFER_SIZE = 8 * 1024 * 2024;
+
+    public StreamCompletionListener(
+        int partNumber,
+        InputStreamContainer blobPartStreamContainer,
+        Path fileLocation,
+        AtomicBoolean anyPartStreamFailed,
+        FileCompletionListener fileCompletionListener
+    ) {
+        this.partNumber = partNumber;
+        this.blobPartStreamContainer = blobPartStreamContainer;
+        this.fileLocation = fileLocation;
+        this.anyPartStreamFailed = anyPartStreamFailed;
+        this.fileCompletionListener = fileCompletionListener;
+    }
+
+    @Override
+    public void onResponse(Void unused) {
+        // Ensures no writes to the file if any stream fails.
+        if (!anyPartStreamFailed.get()) {
+            try (FileChannel outputFileChannel = FileChannel.open(fileLocation, StandardOpenOption.WRITE, StandardOpenOption.CREATE)) {
+                try (InputStream inputStream = blobPartStreamContainer.getInputStream()) {
+                    outputFileChannel.position(blobPartStreamContainer.getOffset());
```

Comment:
Can you use [this helper method in Channels](https://github.com/opensearch-project/OpenSearch/blob/ad7ce4cd446523d52286530ccfbb1dabc2dc7f86/server/src/main/java/org/opensearch/common/io/Channels.java#L243) to simplify some of this code?